jobs/jobs.go

```diff
@@ -1,74 +1,64 @@
-// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
-// See License.txt for license information.
-
 package jobs
 
 import (
-	"sync"
+	"context"
+	"time"
 
 	l4g "github.com/alecthomas/log4go"
-	ejobs "github.com/mattermost/platform/einterfaces/jobs"
 	"github.com/mattermost/platform/model"
 	"github.com/mattermost/platform/store"
-	"github.com/mattermost/platform/utils"
 )
 
-type Jobs struct {
-	startOnce sync.Once
-
-	DataRetention model.Job
-	// SearchIndexing model.Job
-
-	listenerId string
-}
-
-func InitJobs(s store.Store) *Jobs {
-	jobs := &Jobs{
-	// 	SearchIndexing: MakeTestJob(s, "SearchIndexing"),
+func ClaimJob(store store.Store, job *model.Job) bool {
+	// Attempt to claim this job in the database, returning true if successful.
+	// TODO: Implement me properly!
+	if result := <-store.JobStatus().UpdateStatus(job.Id, model.JOB_STATUS_IN_PROGRESS); result.Err != nil {
+		l4g.Error("Failed to claim job: %v", result.Err.Error())
+		return false
 	}
 
-	if dataRetentionInterface := ejobs.GetDataRetentionInterface(); dataRetentionInterface != nil {
-		jobs.DataRetention = dataRetentionInterface.MakeJob(s)
-	}
-
-	return jobs
+	return true
 }
 
-func (jobs *Jobs) Start() *Jobs {
-	l4g.Info("Starting jobs")
-
-	jobs.startOnce.Do(func() {
-		if jobs.DataRetention != nil && *utils.Cfg.DataRetentionSettings.Enable {
-			go jobs.DataRetention.Run()
-		}
-
-		// go jobs.SearchIndexing.Run()
-	})
-
-	jobs.listenerId = utils.AddConfigListener(jobs.handleConfigChange)
+func SetJobProgress(store store.Store, job *model.Job) bool {
+	// TODO: Set the "last updated at" time of the job status.
+	return true
+}
 
-	return jobs
+func SetJobSucceeded(store store.Store, job *model.Job) bool {
+	// TODO: Set the job status to "succeeded".
+	return true
 }
 
-func (jobs *Jobs) handleConfigChange(oldConfig *model.Config, newConfig *model.Config) {
-	if jobs.DataRetention != nil {
-		if !*oldConfig.DataRetentionSettings.Enable && *newConfig.DataRetentionSettings.Enable {
-			go jobs.DataRetention.Run()
-		} else if *oldConfig.DataRetentionSettings.Enable && !*newConfig.DataRetentionSettings.Enable {
-			jobs.DataRetention.Stop()
-		}
-	}
+func SetJobFailed(store store.Store, job *model.Job) bool {
+	// TODO: Set the job status to "failed" and provide details of the error for display in the UI.
+	return true
 }
 
-func (jobs *Jobs) Stop() *Jobs {
-	utils.RemoveConfigListener(jobs.listenerId)
+func SetJobCanceled(store store.Store, job *model.Job) bool {
+	// TODO: Set the job status to "canceled".
+	return true
+}
 
-	if jobs.DataRetention != nil && *utils.Cfg.DataRetentionSettings.Enable {
-		jobs.DataRetention.Stop()
+func CancellationWatcher(ctx context.Context, jobId string, store store.Store, cancelChan chan interface{}) {
+	l4g.Debug("")
+	running := true
+	for running {
```

Comment:
Each worker can only carry out one job at a time, so it'll load the first job of it's type, and a get started on that, along with the cancellation watcher for it.

When shutting down, any jobs that are in progress get canceled currently. Not 100% sure that's desired behaviour though.